substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/resources/NativeImageResourceFileSystem.java

```diff
@@ -0,0 +1,1217 @@
+            checkParents(dir);
+            Entry e = new Entry(dir, true);
+            update(e);
+        } finally {
+            endWrite();
+        }
+    }
+
+    boolean deleteFile(byte[] path, boolean failIfNotExists) throws IOException {
+        IndexNode inode = getInode(path);
+        if (inode == null) {
+            if (path.length == 0) {
+                throw new NativeImageResourceFileSystemException("Root directory </> can't be deleted!");
+            }
+            if (failIfNotExists) {
+                throw new NoSuchFileException(getString(path));
+            } else {
+                return false;
+            }
+        } else {
+            if (inode.isDir() && inode.child != null) {
+                throw new DirectoryNotEmptyException(getString(path));
+            }
+            updateDelete(inode);
+        }
+        return true;
+    }
+
+    private Path createTempFileInSameDirectoryAs() throws IOException {
+        Path tmpPath = Files.createTempFile("rfs_tmp", null);
+        tmpPaths.add(tmpPath);
+        return tmpPath;
+    }
+
+    private Path getTempPathForEntry(byte[] path) throws IOException {
+        Path tmpPath = createTempFileInSameDirectoryAs();
+        if (path != null) {
+            Entry e = getEntry(path);
+            if (e != null) {
+                try (InputStream is = newInputStream(path)) {
+                    Files.copy(is, tmpPath, REPLACE_EXISTING);
+                }
+            }
+        }
+
+        return tmpPath;
+    }
+
+    private void removeTempPathForEntry(Path path) throws IOException {
+        Files.delete(path);
+        tmpPaths.remove(path);
+    }
+
+    void copyFile(boolean deleteSource, byte[] src, byte[] dst, CopyOption[] options) throws IOException {
+        if (Arrays.equals(src, dst)) {
+            return;    // Do nothing, src and dst are the same.
+        }
+
+        beginWrite();
+        try {
+            ensureOpen();
+            Entry eSrc = getEntry(src);  // ensureOpen checked
+
+            if (eSrc == null) {
+                throw new NoSuchFileException(getString(src));
+            }
+            if (eSrc.isDir()) {    // spec says to create dst dir
+                createDirectory(dst);
+                return;
+            }
+            boolean hasReplace = false;
+            boolean hasCopyAttrs = false;
+            for (CopyOption opt : options) {
+                if (opt == REPLACE_EXISTING) {
+                    hasReplace = true;
+                } else if (opt == COPY_ATTRIBUTES) {
+                    hasCopyAttrs = true;
+                }
+            }
+            Entry eDst = getEntry(dst);
+            if (eDst != null) {
+                if (!hasReplace) {
+                    throw new FileAlreadyExistsException(getString(dst));
+                }
+            } else {
+                checkParents(dst);
+            }
+            Entry target = new Entry(eSrc, Entry.COPY);  // Copy eSrc entry.
```

Comment:
When we do a copy of the entry, we may either create a copy of the original entry's bytes, or, if the previous entry is getting deleted, take the original entry's bytes directly. In both cases, however, `copyOnWrite` remains false - if we attempt to write data to this entry (using `getBytes`), it seems like we will accidentally reload the entry - should we also flip the `copyOnWrite` flag to true here?

Reply:
Nice catch. That could cause a bug in the future.